consensus/proto_array/src/proto_array.rs

```diff
@@ -142,11 +125,24 @@ impl ProtoArray {
                 continue;
             }
 
-            let node_delta = deltas
+            let mut node_delta = deltas
                 .get(node_index)
                 .copied()
                 .ok_or(Error::InvalidNodeDelta(node_index))?;
 
+            if self.previous_proposer_boost.root == node.root {
+                node_delta -= self.previous_proposer_boost.score as i64;
+            }
+            if proposer_boost_root != Hash256::zero() && proposer_boost_root == node.root {
+                let num_validators = new_balances.len() as u64;
```

Comment:
I'm looking at some other fork choice code and I noticed that the balances arrays in fork choice can contain entries for inactive validators, which would make this calculation incorrect:

https://github.com/sigp/lighthouse/blob/fff01b24ddedcd54486e374460855ca20d3dd232/beacon_node/beacon_chain/src/beacon_fork_choice_store.rs#L39-L55

I'm not _certain_ that the balances array passed in here comes from that function, but wanted to flag that as something we should check

Reply:
Yes, you're right and we do use that balances array here. It's set in the fork choice store in `set_justified_checkpoint`. Nice catch!

But it also looks like when there's a cache miss in `set_justified_checkpoint`... we are getting balances from `state.balances` where we have exact balances. And I don't think we check whether validators are active at that point. So there seem to be a couple issues here :sweat:

Reply:
Ooft! Ok!

I think once we have the correct effective balances here we _should_ be able to just filter out the 0 balances? So the number of active validators would be `new_balances.iter().filter(|b| b != 0).count()` and the total active balance would be `new_balances.iter().filter(|b| b != 0).sum()`. If we wanted to be super efficient we could compute the sum and the count in one iteration using a `fold` (or a loop).

Reply:
So I guess this is a question of whether we can have active validators with an effective balance of zero? I don't think it's currently possible and it doesn't make sense intuitively, so I'm good with this route.